src/js/view/controls/components/menu/menu.js

```diff
@@ -0,0 +1,499 @@
+                next.focus();
+            }
+        };
+        let childMenu;
+        switch (key) {
+            case 'Esc':
+                settingsMenu.close(evt);
+                break;
+            case 'Left':
+                onLeft();
+                break;
+            case 'Right':
+                onRight();
+                break;
+            case 'Tab': 
+                if (sourceEvent.shiftKey) {
+                    onLeft(true);
+                }
+                break;
+            case 'Up':
+            case 'Down':
+                childMenu = settingsMenu.children[target.getAttribute('name')];
+                if (childMenu) {
+                    childMenu.open(evt);
+                    if (childMenu.items && childMenu.items.length) {
+                        childMenu.items[0].el.focus();
+                    }
+                }
+                break;
+            default:
+                break;
+        }
+        sourceEvent.stopPropagation();
+        if (/13|32|37|38|39|40/.test(sourceEvent.keyCode)) {
+            // Prevent keypresses from scrolling the screen
+            sourceEvent.preventDefault();
+            return false;
+        }
+    });
+    return ui;
+};
+
+export const SettingsMenu = (api, model, controlbar, localization) => {
+    const settingsMenu = new Menu('settings', null, localization);
+
+    const changeMenuItems = (menuName, items, onItemSelect, defaultItemIndex, itemOptions) => {
+        const controlBarButton = controlbar.elements[`${menuName}Button`];
+        if (!items || items.length <= 1) {
+            settingsMenu.removeMenu(menuName);
+            if (controlBarButton) {
+                controlBarButton.hide();
+            }
+            return;
+        }
+        let menu = settingsMenu.children[menuName];
+        if (!menu) {
+            menu = new Menu(menuName, settingsMenu, localization);
+        }
+        menu.setMenuItems(
+            menu.createItems(items, onItemSelect, itemOptions), 
+            defaultItemIndex
+        );
+        if (controlBarButton) {
+            controlBarButton.show();
+        }
+    };
+    
+    model.change('levels', (changedModel, levels) => {
+        const menuItemOptions = { defaultText: localization.auto };
+        changeMenuItems('quality', levels, api.setCurrentQuality, model.get('currentLevel'), menuItemOptions);
+        const childMenus = settingsMenu.children;
+        const shouldShowGear = !!childMenus.quality || !!childMenus.playbackRates || Object.keys(childMenus).length > 1;
+        controlbar.elements.settingsButton.toggle(shouldShowGear);
+    }, settingsMenu);
+    model.change('captionsList', (changedModel, captionsList) => {
+        const menuItemOptions = { defaultText: localization.off };
+        const initialIndex = model.get('captionsIndex');
+        changeMenuItems('captions', captionsList, api.setCurrentCaptions, initialIndex, menuItemOptions);
```

Comment:
i like this pattern a lot 🎉

Reply:
There goes my feeble attempt to get around line length limitations x)